src/uct/ib/rc/accel/rc_mlx5_ep.c

```diff
@@ -144,6 +145,141 @@ ucs_status_t uct_rc_mlx5_ep_put_short(uct_ep_h tl_ep, const void *buffer, unsign
+    }
+}
+
+static ucs_status_t UCS_F_ALWAYS_INLINE
+uct_rc_mlx5_ep_short_dm(uct_rc_mlx5_ep_t *ep, uct_rc_mlx5_dm_copy_data_t *cache,
+                        size_t hdr_len, const void *payload, unsigned length,
+                        unsigned opcode, uint8_t fm_ce_se,
+                        uint64_t rdma_raddr, uct_rkey_t rdma_rkey)
+{
+    uct_rc_mlx5_iface_t *iface = ucs_derived_of(ep->super.super.super.iface, uct_rc_mlx5_iface_t);
+    uct_rc_iface_t *rc_iface   = &iface->super;
+    uct_rc_iface_send_desc_t *desc;
+    void *buffer;
+
+    desc = ucs_mpool_get_inline(&iface->mlx5_common.dm.dm->mp);
+    if (ucs_unlikely(desc == NULL)) {
+        /* in case if no resources available - fallback to bcopy */
+        UCT_RC_IFACE_GET_TX_DESC(rc_iface, &rc_iface->tx.mp, desc);
+        desc->super.handler = (uct_rc_send_handler_t)ucs_mpool_put;
+        buffer = desc + 1;
+
+        /* condition is static-evaluated, no performance penalty */
+        if (cache && hdr_len) {
+            memcpy(desc + 1, cache->out, hdr_len);
+        }
+        memcpy(UCS_PTR_BYTE_OFFSET(desc + 1, hdr_len), payload, length);
+    } else {
+        ucs_assert(desc->super.buffer != NULL);
+        buffer = (void*)(desc->super.buffer - iface->mlx5_common.dm.dm->start_va);
+
+        uct_rc_mlx5_ep_copy_to_dm(cache, hdr_len, payload,
+                                  length, desc->super.buffer);
+
+    }
+    uct_rc_mlx5_txqp_bcopy_post(rc_iface, &ep->super.txqp, &ep->tx.wq,
+                                opcode, hdr_len + length,
+                                rdma_raddr, rdma_rkey, fm_ce_se,
+                                0, desc, buffer);
+    return UCS_OK;
+}
+#endif
+
+ucs_status_t
+uct_rc_mlx5_ep_put_short(uct_ep_h tl_ep, const void *buffer, unsigned length,
+                         uint64_t remote_addr, uct_rkey_t rkey)
+{
+#if HAVE_IBV_EXP_DM
+    uct_rc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_rc_mlx5_iface_t);
+    uct_rc_iface_t *rc_iface   = &iface->super;
+    uct_rc_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_rc_mlx5_ep_t);
+    ucs_status_t status;
+
+    if (ucs_likely((length <= UCT_IB_MLX5_PUT_MAX_SHORT(0)) ||
+                   !iface->mlx5_common.dm.dm)) {
+        return uct_rc_mlx5_ep_put_short_inline(tl_ep, buffer, length, remote_addr, rkey);
+    }
+
+    UCT_CHECK_LENGTH(length, 0, iface->mlx5_common.dm.seg_len, "put_short");
+    UCT_RC_CHECK_RES(rc_iface, &ep->super);
+    status =  uct_rc_mlx5_ep_short_dm(ep, NULL, 0, buffer, length,
+                                      MLX5_OPCODE_RDMA_WRITE,
+                                      MLX5_WQE_CTRL_CQ_UPDATE,
+                                      remote_addr, rkey);
+    if (UCS_STATUS_IS_ERR(status)) {
+        return status;
+    }
+    UCT_TL_EP_STAT_OP(&ep->super.super, PUT, SHORT, length);
+    return UCS_OK;
+#else
+    return uct_rc_mlx5_ep_put_short_inline(tl_ep, buffer, length, remote_addr, rkey);
```

Comment:
can we rearrange the if/#ifdef so that uct_rc_mlx5_ep_put_short_inline() would be used only once and not twice?